Hands-On Lab: Building a Three-Way Split    
● Step 1: Take a Week 3 dataset and create a 60/20/20 train/validation/test split with a fixed random_state.    
● Step 2: Train a model on the training set and tune one setting by checking the validation set only.   
● Step 3: Evaluate the final model on the test set exactly once and report the score.   
● Step 4: In Markdown, explain what would go wrong if you had tuned against the test set instead.   

Taking week3/day3 dataset (breast cancer).

In [10]:
from sklearn.datasets import load_breast_cancer

from sklearn.model_selection import train_test_split

from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import accuracy_score , roc_auc_score, classification_report

loading & splitting the dataset:

In [11]:
Br_Can = load_breast_cancer() 

X = Br_Can.data


y = (Br_Can.target == 0).astype(int) # make it into: malignant = 1, benign = 0


# To achive a 60/20/20 split, at first split data into temp 80% and test 20% , then take the temp 80% and split it into train 75% and validation 25% 

X_temp , X_test , y_temp, y_test = train_test_split( X , y , test_size=0.2 , random_state= 42 ) 

X_train, X_val, y_train, y_val = train_test_split(X_temp, y_temp, test_size= 0.25, random_state= 42)



fitting and testing the model on the validation part:

In [12]:
RF = make_pipeline(StandardScaler(), RandomForestClassifier())

RF.fit(X_train, y_train)
Val_prediction = RF.predict(X_val)


Val_pred_proba = RF.predict_proba(X_val)

Val_acc= accuracy_score(y_val, Val_prediction)

Val_Roc = roc_auc_score(y_val, Val_pred_proba[:,1] ) # calculate ROC using predicted probabilities for the positive class(maliginant)

print(f"Accuracy achived using validation part: {Val_acc : .3f}")
print(f"ROC score from Validation part: {Val_Roc : .3f}")


Accuracy achived using validation part:  0.947
ROC score from Validation part:  0.978


fitting and testing the model on the Testing part:

In [13]:
Test_prediction = RF.predict(X_test)

Test_pred_proba = RF.predict_proba(X_test)

Test_acc= accuracy_score(y_test, Test_prediction)

Test_Roc = roc_auc_score(y_test, Test_pred_proba[:,1] ) # calculate ROC using predicted probabilities for the positive class(maliginant)

print(f"Accuracy achived using Testing part: {Test_acc : .3f}")
print(f"ROC score from Testing part: {Test_Roc : .3f}")

Accuracy achived using Testing part:  0.956
ROC score from Testing part:  0.996


Accuracy achived using validation part:  0.947  
ROC score from Validation part:  0.978  

Accuracy achived using Testing part:  0.956    
ROC score from Testing part:  0.996     

Since there isn't a big variance in results, these outputs verify that our model is as accurate on new data as in our testing.

Tuning againest the test set will not yeild an honest/accurate estimate of the model's performance on new unexplored data.  